pkg/api/validation/validation.go

```diff
@@ -2571,7 +2571,9 @@ func validateServicePort(sp *api.ServicePort, requireName, isHeadlessService boo
 func ValidateServiceUpdate(service, oldService *api.Service) field.ErrorList {
 	allErrs := ValidateObjectMetaUpdate(&service.ObjectMeta, &oldService.ObjectMeta, field.NewPath("metadata"))
 
-	if api.IsServiceIPSet(oldService) {
+	// ClusterIP should be mutable only for empty value, to allow assigning ClusterIP
+	// for service changing type to one that requires ClusterIP
+	if oldService.Spec.ClusterIP != "" {
```

Comment:
I wonder if it makes more sense to refactor this into a top-level switch on service.Spec.Type and just handle each case in a more obvious way...  Thoughts?